### PR TITLE
Add AMI command for aws code generation

### PR DESCRIPTION
**Pull request by MarcelMue**
Tested locally with `./devctl gen ami --dir .` which produces the wonderful `ami.go` we are used to.

### CODE REVIEW

**Comment by tfussell** [comment]
I just tested this and it works well. It seems to run twice for some reason though:
```
devctl gen ami --dir service/controller/key
scraping https://stable.release.flatcar-linux.net/amd64-usr/
scraping https://stable.release.flatcar-linux.net/amd64-usr/2191.5.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2247.5.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2247.6.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2247.7.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2303.3.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2303.3.1/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2303.4.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2345.3.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2345.3.1/flatcar_production_ami_all.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2247.6.0.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2303.3.0.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2303.3.1.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2345.3.0.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2191.5.0.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2247.7.0.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2303.4.0.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2345.3.1.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2247.5.0.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/
scraping https://stable.release.flatcar-linux.net/amd64-usr/2191.5.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2247.5.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2247.6.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2247.7.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2303.3.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2303.3.1/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2303.4.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2345.3.0/flatcar_production_ami_all.json
scraping https://stable.release.flatcar-linux.net/amd64-usr/2345.3.1/flatcar_production_ami_all.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2191.5.0.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2247.6.0.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2303.3.0.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2345.3.0.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2345.3.1.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2247.5.0.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2247.7.0.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2303.3.1.json
scraping https://flatcar-prod-ami-import-cn-north-1.s3.cn-north-1.amazonaws.com.cn/stable/amd64-usr/2303.4.0.json
```

File: pkg/gen/ami/ami.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+	"strings"
+
+	"github.com/Masterminds/semver"
+	"github.com/giantswarm/microerror"
+	"golang.org/x/net/html"
+
+	"github.com/giantswarm/devctl/pkg/gen/input"
+)
+
+var amiTemplate = `// NOTE: This file is generated. Do not edit.
```
Comment (tfussell) [inline pkg/gen/ami/ami.go:20]:
```suggestion
var amiTemplate = `// NOTE: This file is generated by devctl. Do not edit.
```

File: pkg/gen/ami/ami.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"strings"
+
+	"github.com/Masterminds/semver"
```
Comment (tfussell) [inline pkg/gen/ami/ami.go:13]:
```suggestion
	"github.com/Masterminds/semver/v3"
```

Comment (tfussell) [reply pkg/gen/ami/ami.go:13]: We should use semver v3 since that's the go modules version. You'll need to use `go mod tidy` after this change.

**Comment by MarcelMue** [comment]
I also verified that it does the correct thing by checking in `aws-operator` with command:
```
devctl gen ami --dir ./service/controller/key/
```

File: pkg/gen/ami/ami.go (hunk 2 above)
Comment (MarcelMue) [reply pkg/gen/ami/ami.go]: Did this locally to clean right away.

**Comment by MarcelMue** [comment]
I checked why it is executed twice:
The underlying templating implementation in devctl asks for inputs twice for one render.
Once here: https://github.com/giantswarm/devctl/blob/97d1f08c97733ca7c209fb2d23b6c137bda54709/pkg/gen/gen.go#L16
And again here:
https://github.com/giantswarm/devctl/blob/97d1f08c97733ca7c209fb2d23b6c137bda54709/pkg/gen/internal/gen.go#L15

It looks like @kopiczko and @tuommaki implemented each function but both are called on every render. So my assumption must be that it is a mistake.

I would vouch to fix this in a follow up.

File: pkg/gen/ami/ami.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+	templateData := map[string]interface{}{
+		"AMIInfoString": fmt.Sprintf("`%s`", result),
+		"Package":       "key",
+	}
+
+	return templateData, nil
+}
+
+func scrapeVersions(source io.Reader) ([]string, error) {
```
Comment (kopiczko) [inline pkg/gen/ami/ami.go:155]: This is some serious shit.

File: pkg/gen/ami/ami.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+	"strings"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/giantswarm/microerror"
+	"golang.org/x/net/html"
+
+	"github.com/giantswarm/devctl/pkg/gen/input"
+)
+
+var amiTemplate = `// NOTE: This file is generated by devctl. Do not edit.
```
Comment (kopiczko) [inline pkg/gen/ami/ami.go:20]: I'd like to move it to the bottom of the file.

**Comment by tfussell** [comment]
> I checked why it is executed twice:
> The underlying templating implementation in devctl asks for inputs twice for one render.
> Once here:
> 
> https://github.com/giantswarm/devctl/blob/97d1f08c97733ca7c209fb2d23b6c137bda54709/pkg/gen/gen.go#L16
> 
> 
> And again here:
> https://github.com/giantswarm/devctl/blob/97d1f08c97733ca7c209fb2d23b6c137bda54709/pkg/gen/internal/gen.go#L15
> 
> It looks like @kopiczko and @tuommaki implemented each function but both are called on every render. So my assumption must be that it is a mistake.
> 
> I would vouch to fix this in a follow up.

Makes sense. It's not that important.

File: cmd/gen/ami/runner.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		return microerror.Mask(err)
+	}
+
+	err = gen.Execute(
+		ctx,
+		amiFile,
+	)
+	if gen.IsFilePath(err) {
+		fmt.Fprintf(r.stderr, "%s\n", err)
+		os.Exit(1)
```
Comment (xh3b4sd) [inline cmd/gen/ami/runner.go:61]: I think we should just do proper error handling and do the fancy UI stuff as close to the root of program initialization as possible. IMO it is an antipattern to kill the program somewhere in the middle of the code.

Comment (MarcelMue) [reply cmd/gen/ami/runner.go]: Removed - other commands in devctl still do this (where I aligned) so we might want to clean that up in the future.

File: pkg/gen/ami/ami.go (hunk 4 above)
Comment (MarcelMue) [reply pkg/gen/ami/ami.go]: Done.

File: pkg/gen/ami/ami.go (hunk 3 above)
Comment (tfussell) [reply pkg/gen/ami/ami.go:139]: I always knew my Python HTML scraping experience from college would be useful one day.

**Pull request merged**